src/main/java/build/buildfarm/worker/AbstractPipelineStage.java

```diff
@@ -0,0 +1,116 @@
+// Copyright 2017 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ArrayBlockingQueue;
+
+abstract class AbstractPipelineStage implements PipelineStage {
+  private final PipelineStage output;
+  private final PipelineStage error;
+  private final BlockingQueue<OperationContext> queue;
+
+  private PipelineStage input;
```

Comment:
It looks like you need all these variables in subclasses, so maybe add protected accessors?

Reply:
There's no usage of these, but that highlights that these are completely unnecessary except for Executor->ExecuteActionStage communication.  I've updated the diff to clean this up.